### PR TITLE
Expand Windows EXE detection regex

### PR DESCRIPTION
**Pull request by johnwparent**
.bat or .exe files can be considered executable on Windows. This PR expands the regex for detectable packages to allow for the detection of packages that vendor .bat wrappers (intel mpi for example)
This approach has the added benefit of inherently supporting the mocked testing suite executables (which are .bat files)

### CODE REVIEW

**Comment by johnwparent** [comment]
@alalazo this should resolve the monkeypatch issues you're having with #39843

**Review by scheibelp — DISMISSED** [review]
This looks like a good improvement (since otherwise, a package that lists `executables` would not be detected if they were `.bat`). I don't understand the side effect you describe at https://github.com/spack/spack/pull/39850#issuecomment-1708836707 though all of our monkeypatches should be preserved between parent and child process.

I'm inclined to merge this as-is, but @alalazo did you want to try this out before that?

File: lib/spack/spack/detection/path.py
Comment (alalazo) [inline lib/spack/spack/detection/path.py]: Is there any reason to add this conditional?

**Review by alalazo — COMMENTED** [review]
I just have a question but otherwise LGTM. Feel free to merge and I'll rebase #39843

File: lib/spack/spack/detection/path.py
Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: Otherwise we search non specified paths when `path_hints` is given. By this point, `path_hints` is always truthy, but those extra paths are package dependent, so they need to be invoked at that point, but querying `path_hints` is meaningless. The conditional stores whether `path_hints` was provided and then prevents the extra search paths from being considered if so.
The prior behavior was incorrect.

File: lib/spack/spack/test/cmd/external.py
Comment (scheibelp) [inline lib/spack/spack/test/cmd/external.py]: Two issues:

* I think all tests running `external("find")` in this module would be safer using this.
* (not necessarily crucial here but possibly meriting a TODO comment about refactoring) The piecemeal monkeypatching of individual path aggregation actions suggests to me that a refactor would be useful.

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: An empty list should also mean no path hints so I think `no_path_hints = not bool(path_hints)` would be better.

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: I think this line could be rewritten more clearly as:

```
if not path_hints:
  path_hints = spack.util.environment.get_path("PATH")
```

and actually it looks like regardless of whether we pull from `PATH`, if `path_hints` was originally empty, then we include  `common_windows_package_paths`, so we can also fold that in like:

```
if not path_hints:
  path_hints = spack.util.environment.get_path("PATH")
  path_hints.extend(common_windows_package_paths())
```

**Review by scheibelp — CHANGES_REQUESTED** [review]

File: lib/spack/spack/detection/path.py
Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: `path_hints is None` was a deliberate choice to avoid casting `path_hints` to bool as this mimic the behavior of the next line, which also uses `None` as opposed to booleans, and in doing so, ensures behavior dependent on `no_path_hints` remains the same as if it were part of the following command, which was the intent behind `no_path_hints`.
I assumed the choice below to explicitly use `None` was a deliberate one. If this is not the case, then I'll refactor both. Which ever way we decide to go, these two lines should use the same check. I only don't refer to `no_path_hints` in the following line because `mypy` can't seem to handle it and I've already spent more time than I'd like trying to reconcile that.

File: lib/spack/spack/detection/path.py
Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: Outside of the concern re the conditional I raised above, this looks good.

File: lib/spack/spack/test/cmd/external.py
Comment (johnwparent) [reply lib/spack/spack/test/cmd/external.py]: Agreed on the first statement, however I think the second point here warrants further consideration. In general, tests in this module that care about isolation already are arbitrarily monkeypatching the PATH to point to some location containing only the executables they want the test to detect. However, on Windows, we don't only search the PATH, we search several other common installation locations that are used by other tools like VS, choco, or vcpkg to name a few. So in addition to monkey patching the PATH, on Windows, to achieve the degree of isolation expected by the tests, we must monkey patch further. This is the equivalent of _many_ calls to monkeypatch various environment variables, and is essentially no different from what is being done, except this is more concise.  Each of these methods is singularly responsible for aggregating a conceptually distinct set of potential package locations, and two of the methods aggregate locations that are dependent on the package being searched, whereas the third is just a common set of locations universal to all packages.  We could group the other two, no issue there, but I just want to emphasize that there is a deliberate reason this is setup this way, and that it's not just the result of iterative development missing the potential for a consolidated interface.

File: lib/spack/spack/detection/path.py
Comment (alalazo) [reply lib/spack/spack/detection/path.py]: I believe the intent is to register if the user passed anything as argument, instead of using the default value. In that sense:
```
path_hints is None
```
is the correct choice. In particular in the line below, this allows distinguishing among "default value" (for which we look into the PATH env variable), and an empty list that was passed explicitly.

File: lib/spack/spack/detection/path.py
Comment (alalazo) [reply lib/spack/spack/detection/path.py]: This line:
```
path_hints = spack.util.environment.get_path("PATH") if path_hints is None else path_hints
```
should distinguish between user not passing values, like:
```
detection.by_path(["gcc", "cmake"])
```
in which case we inspect the PATH, and:
```
detection.by_path(["gcc", "cmake"], path_hints=[])
```
in which case we don't search anything (since the hint is explicitly empty).

Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: Ok, I thought that might have been deliberate. I think we can still make this check:
```
if no_path_hints:
  path_hints = spack.util.environment.get_path("PATH")
  path_hints.extend(common_windows_package_paths())
```
and just make the `path_hints is None` check once.

Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: Ok, seems like mypy is still unable to follow control flow so these checks will now be:
```
    if path_hints is None:
        path_hints = spack.util.environment.get_path("PATH") if path_hints is None else path_hints
        path_hints.extend(common_windows_package_paths())
        no_path_hints = True
```
Which I think achieves what everyone is looking for here.

Comment (scheibelp) [reply lib/spack/spack/detection/path.py]: @alalazo

> detection.by_executable(["gcc", "cmake"], path_hints=[])
> in which case we don't search anything (since the hint is explicitly empty).

I like to avoid that type of logic whenever possible: forcing users to distinguish between empty/None in Python is a source of confusion. If we want to be able to disable defaults in the presence of `[]`, then IMO we should have a `disable_defaults` parameter. However, my assessment is that we wouldn't want that, because a combined `path_hints=[], disable_defaults=True)` would lead to an empty set of search paths on Linux (i.e. we don't look in any paths, so we shouldn't be calling this function in the first place).

I can say that none of the call sites in core that I can find depend on this distinction:

```
lib/spack/spack/cmd/external.py:    detected_packages = spack.detection.by_executable(pkg_cls_to_check, path_hints=args.path)
lib/spack/spack/bootstrap/core.py:    externals = spack.detection.by_executable(search_list)
lib/spack/spack/build_systems/python.py:        python_externals_detection = spack.detection.by_executable(
```

**Comment by johnwparent** [comment]
@scheibelp @alalazo Post rebase with Max's PR I had to make quite a few changes to disable the default Windows search paths if path hints is none (which will require minimal changes if @scheibelp's suggestion is the route we take), so you both may want to take a closer look than would be normal for this point in review.

File: lib/spack/spack/detection/path.py
Comment (alalazo) [reply lib/spack/spack/detection/path.py]: > I like to avoid that type of logic whenever possible: forcing users to distinguish between empty/None in Python is a source of confusion.

Hmm, I think it is a common idiom used in many places - also in the standard library. This is from the ProcessPoolExecutor constructor:
```python
        if max_workers is None:
            self._max_workers = os.cpu_count() or 1
            if sys.platform == 'win32':
                self._max_workers = min(_MAX_WINDOWS_WORKERS,
                                        self._max_workers)
        else:
            if max_workers <= 0:
                raise ValueError("max_workers must be greater than 0")

```
where they distinguish between `0` (not allowed) and `None`. The idiom is so common that `mypy` has an `Optional[X]` type hint that means the value might be type `X` or `None` , see [here](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html).

> I can say that none of the call sites in core that I can find depend on this distinction:

This in my opinion doesn't matter. We need a way to distinguish when to use a default value from PATH and when the client passed an empty list. Even from a common sense point of view:
```
specs = by_path(["cmake"])  # cmake found in PATH, not unexpected
specs = by_path(["cmake"], path_hints=[])  # cmake found in PATH ? Let's give strictly MORE things to search
specs = by_path(["cmake"], path_hints=["/tmp/foo"])  # cmake not found ? 
```

Comment (scheibelp) [reply lib/spack/spack/detection/path.py]: Note that

`specs = by_path(["cmake"], path_hints=[])`

(with the semantics you desire) will never return anything on Linux, it is effectively a noop (so this type of call should never occur explicitly).

That being said, you are likely referring to cases where you compose `path_hints` programmatically. I'd say this creates confusion when you want to use default paths:

```
computed_path_hints = function() # let's say this is an empty list
# now I need to convert computed_path_hints to None explicitly if I want `by_executable` to find anything
```

I think that putting a `disable_defaults` parameter in the method signature is a clearer way of expressing this.

> I think it is a common idiom used in many places

That does not mean it's the right way to handle it here. I'm not planning a crusade on abuse of null values in our code base, but since I'm here making other requests in the vicinity of this code, I want to change this too.

Comment (alalazo) [reply lib/spack/spack/detection/path.py]: > Note that ... never occur explicitly

No, but they could occur when the value is read from another function:
```python
specs = by_path(["cmake"], path_hints=list_of_paths_to_search())
```

> I think that putting a disable_defaults parameter in the method signature is a clearer way of expressing this.

That's exactly what `None` means idiomatically in Python - I think it is a widely accepted idiom, while it's discouraged to use _falsey_ values[^1], as they can be anything that can be casted to False. What I mean is that I did that change on purpose :slightly_smiling_face: 

[^1]: I can double check for a reference, but iirc at least "Python Cookbook" from D. Beazley has a few lines about this

Comment (scheibelp) [reply lib/spack/spack/detection/path.py]: I feel like this response ignores what I said after

>> (with the semantics you desire) will never return anything on Linux, it is effectively a noop (so this type of call should never occur explicitly).

for example

> No, but they could occur when the value is read from another function:

I specifically anticipated this argument and addressed it.

>> I think that putting a disable_defaults parameter in the method signature is a clearer way of expressing this.

> That's exactly what None means idiomatically in Python

Yes, and I'm not a fan (the example I gave in https://github.com/spack/spack/pull/39850#discussion_r1321824504 demonstrates why). I'm worried about abuse of null; you are worried (possibly) about abuse of false-yness; my proposal to use `disable_defaults` addresses both (or at least, my proposal does not *require* abuse of false-yness).

Comment (alalazo) [reply lib/spack/spack/detection/path.py]: > Yes, and I'm not a fan (the example I gave in https://github.com/spack/spack/pull/39850#discussion_r1321824504 demonstrates why).

I don't think this example demonstrates it, as:
```
# Now I need to convert computed_path_hints to None explicitly if I want `by_executable` to find anything
```
implies that a user expects something to happen if an empty list is passed explicitly, which should not be the case. 

> my proposal to use disable_defaults addresses both (or at least, my proposal does not require abuse of false-yness).

In my opinion this adds an unnecessary argument, but I don't want to hold this PR on this. I think we just need to add a line in the docstring, to say:
```
If path_hints is None the paths are computed using defaults depending on the current platform.
```
or something similar.

Comment (alalazo) [reply lib/spack/spack/detection/path.py]: As an additional point, this is the code before #39843:

https://github.com/spack/spack/blob/7631b5ea14fb48f8859015b48fc379e06c9d4141/lib/spack/spack/detection/path.py#L233-L245

This behavior (which I think is correct) was introduced back in #33021 Before we had the _falsey_ kind of check.  

Proposal: can't we keep the "optional" behavior as it is in this PR and debate the internal API in another PR ?

Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: > Proposal: can't we keep the "optional" behavior as it is in this PR and debate the internal API in another PR ?

Sorry to chime in, but FWIW I agree this is somewhat outside the scope of this PR and seems to be a higher level debate than this PR was meant to be. If we could get this merged (both so we can re-enable those skipped tests and to support `Intel-MPI` detection on Windows (compiler wrapper detection via .bat)) and then I can make another PR to continue this conversation?

File: lib/spack/spack/test/cmd/external.py
Comment (scheibelp) [reply lib/spack/spack/test/cmd/external.py]: > Agreed on the first statement, 

OK, there are several tests in this module that call `spack external find` and don't use this fixture at the moment.

> So in addition to monkey patching the PATH, on Windows, to achieve the degree of isolation expected by the tests, we must monkey patch further. This is the equivalent of many calls to monkeypatch various environment variables, and is essentially no different from what is being done, except this is more concise.

What I was suggesting, was that the code should be refactored such that the point that pulls in default paths for Windows should call in a single function, which outside of these tests, should immediately call out to the 3 or so functions you monkeypatch; in the tests, we should monkeypatch just that one function. If we add another default search location on Windows, it won't suddenly appear in our tests.

Comment (scheibelp) [reply lib/spack/spack/test/cmd/external.py]: >> Agreed on the first statement,

> OK, there are several tests in this module that call spack external find and don't use this fixture at the moment.

Whoops I see the fixture was promoted to module scope

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: Our docs builder doesn't like the `*` (it thinks you are trying to bold something) - I recommend just saying Unix

**Comment by scheibelp** [comment]
I retract all requests related to default interpretations of `None`. That being said: `test_packages_yaml_format ` is failing in the Windows unit tests, and there is a docs error.

**Comment by johnwparent** [comment]
@alalazo I seem to be getting behavior similar to what you had mentioned experiencing on #39843 where it seems like pytest fixtures are not propagating to the threads but are available from the main context. I added a throw in one of the fixtures and nothing happened so seems to indicate that they are indeed being skipped. How did you resolve this? This seems to be source of the test failures.

cc @scheibelp as IIRC he had some thoughts on this

Edit:
For more context:
https://github.com/spack/spack/actions/runs/6162399453
https://github.com/spack/spack/actions/runs/6164456504
are back to back workflows, the only difference between them is a rebase on master and the failures seem to somewhat arbitrary. FWIW I cannot make any of these failures occur locally, but I can make a completely different test fail for seemingly unrelated reasons to this PR.

**Comment by alalazo** [comment]
@johnwparent The failure on the abseil build is due to timeout. It's sad to say, but even with 2 mins. as a timeout threshold sometimes we reach it in CI (2 mins to literally call `cmake --version` and get back the output as string).

The issue with unit tests instead is multiprocessing + spawn context + monkeypatching. The child processes don't see the monkey patching done in the test, because they are "spawned" and not "forked". Depending on the test, we might have different solutions:
1. Avoid monkey-patching in the first place
2. Test the function executed in the child directly, without spawning multiple processes
3. Do something as @scheibelp does and pass the context to the child process

I would go for 1 or 2 if feasible, since they result in simpler tests and less boilerplate - but I have to say that 3 always works (we basically have to write code to ensure monkeypatching is applied in the child process too).

**Comment by alalazo** [comment]
For reference, here's the log in the abseil build that failed:

**Comment by johnwparent** [comment]
> @johnwparent The failure on the abseil build is due to timeout. It's sad to say, but even with 2 mins. as a timeout threshold sometimes we reach it in CI (2 mins to literally call `cmake --version` and get back the output as string).
> 
> The issue with unit tests instead is multiprocessing + spawn context + monkeypatching. The child processes don't see the monkey patching done in the test, because they are "spawned" and not "forked". Depending on the test, we might have different solutions:
> 
> 1. Avoid monkey-patching in the first place
> 2. Test the function executed in the child directly, without spawning multiple processes
> 3. Do something as @scheibelp does and pass the context to the child process
> 
> I would go for 1 or 2 if feasible, since they result in simpler tests and less boilerplate - but I have to say that 3 always works (we basically have to write code to ensure monkeypatching is applied in the child process too).

Thanks for the context @alalazo! 

I completely skimmed over that external search timeout, but even if we don't detect CMake as an external we should really be able to install it, which is failing with this error:
```
Flagging cmake-3.27.4-6tywfczqd6fjpabokslp3ju4rrmowcfa as failed: Link path (D:\a\spack\spack\opt\spack\windows-windows10.0.20348-AMD64\msvc-19.37.32822\cmake-3.27.4-6tywfczqd6fjpabokslp3ju4rrmowcfa\bin\zlib1.dll) already exists. Cannot create link.
```
which I think #39933 should resolve. 

Thanks for the guidance re the multi processing. I thought it might've been some issue with `spawn` not propagating anything not directly reference by the child process. I'll try option 3 as we do need this monkey patch and the end to end externals test can't really work around the need for a child process.

File: lib/spack/spack/detection/path.py
Comment (alalazo) [inline lib/spack/spack/detection/path.py]: @johnwparent Can we find a solution that doesn't require to modify an inner class just for testing? I'm happy to help with that in case.

Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: @scheibelp and I had discussed and landed on this approach (we do this in the install process as well) but I'd be happy to discuss alternatives.

AFAICT the only way to transparently propagate the monkeypatches to a subprocess only during test time would be either to monkeypatch the `Finder` class or modify the subprocess creation/pickling process. Am I missing something?

Comment (alalazo) [reply lib/spack/spack/detection/path.py]: Summarizing a discussion in DMs on Slack. We can try splitting tests on parent / child boundaries so to never have to spawn a process when we use monkey patching. For instance, the:
```
test_packages_yaml_format
```
test might just check that `spack.detection.update_configuration` works correctly for some well-defined input.

File: lib/spack/spack/detection/path.py
Comment (alalazo) [inline lib/spack/spack/detection/path.py]: Can you expand on this? I don't quite understand the problem and how it is solved :thinking:

Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: Sorry, I should have taken this out. This was one of my earlier attempts to pass the monkeypatches into the subprocess. It's since been rendered totally unnecessary and I was planning on dropping it once we ok'd the new test I added and I went back to cleanup the history.

One of my approaches was to abuse the pickling methods for a class to rope in the monkeypatches but I felt like it was too invasive to the core code. This is the last vestige of that approach.

**Comment by johnwparent** [comment]
@scheibelp @alalazo I've cleaned up the history and added the discussed unit test in place of the failing end-end test (same functionality is tested).

File: lib/spack/spack/detection/path.py
Comment (alalazo) [inline lib/spack/spack/detection/path.py:341]: I think here we changed behavior, and now we are using `PATH` as a guess for libraries too. Am I reading this right? 

Apart from this question, all the rest LGTM.

Comment (johnwparent) [reply lib/spack/spack/detection/path.py:341]: I think this would only be a change for non Windows. Windows we've always searched the PATH for both binaries and libraries

Comment (johnwparent) [reply lib/spack/spack/detection/path.py:341]: If we don't want this I'm happy to refactor in a way that only libraries on Windows will search the PATH

Comment (alalazo) [reply lib/spack/spack/detection/path.py:341]: I don't think on linux it makes much sense to search `PATH` directories for libraries. Maybe it would be better to search in things like `LD_LIBRARY_PATH` or similar (obviously, this new search is subject for another PR, not for this one)

Comment (johnwparent) [reply lib/spack/spack/detection/path.py:341]: Ok, I'll prevent the PATH from being searched on unix

Comment (johnwparent) [reply lib/spack/spack/detection/path.py:341]: Addressed

**Comment by johnwparent** [comment]
@scheibelp I finally seem to have fixed that flakey test (or something in develop did) so I believe this is ready for (optimistically) a final round.

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py:379]: Is this ever reached? There is an `except Exception` right above it. I think you'd have to change the above catch to `except concurrent.futures.TimeoutError`.

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py:262]: Can you update the docstring of this method to reflect the new behavior?

(Originally, if `initial_guess = []`, then this would still add default paths via `path_hints`, now it doesn't)

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: This function uses defaults when `path_hints=[]`, which runs counter to the behavior of other functions we've discussed in this PR, as well as the parents - is that intended?

I think docstrings would help clarify the intent.

**Review by scheibelp — CHANGES_REQUESTED** [review]
Docs/exception-handling requests

File: lib/spack/spack/detection/path.py
Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: Addressed and pushed.

File: lib/spack/spack/detection/path.py
Comment (johnwparent) [reply lib/spack/spack/detection/path.py:262]: Addressed and pushed.

File: lib/spack/spack/detection/path.py
Comment (johnwparent) [reply lib/spack/spack/detection/path.py:379]: Yeah, don't know how I messed that up. Fixed and pushed.

**Comment by johnwparent** [comment]
@spackbot fix style

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: Why does this not take `paths`? I would assume that is what would get passed if a user ran `spack external find --path=...`)

Part of the issues is I think from examining candidate files, that `libraries_in_windows_paths(... path_hints=...)` would never be `None`. If that's so though, then I think the interface should be refined

My guess as to how to get sensible behavior is to promote the conditional hint augmentation from that function up to here (or the caller of this function).

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a new question about the propagation of `path_hints=...`

File: lib/spack/spack/detection/path.py
Comment (johnwparent) [reply lib/spack/spack/detection/path.py]: I think this was just a mistake as I was hurrying to make edits, it should be exactly the same as the unix implementation.

**Review by scheibelp — APPROVED** [review]

**Pull request merged**
